### ISSUE TITLE
How was the Server ever supposed to accept multiple connections?

### ISSUE THREAD

**Issue by mrzzzrm** [hyrise/hyrise#1370]
Maybe this is related to #1357, #1032 and #853 but I am opening a separate issue to quickly close and forget because maybe I am also just stupid.

This is the Server:

```c++
Server::Server(boost::asio::io_service& io_service, uint16_t port)
    : _io_service(io_service),
      _acceptor(io_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), port)),
      _socket(io_service) {
  _accept_next_connection();
}

void Server::_accept_next_connection() {
  _acceptor.async_accept(_socket, boost::bind(&Server::_start_session, this, boost::asio::placeholders::error));
}

void Server::_start_session(boost::system::error_code error) {
  if (!error) {
    auto connection = std::make_shared<ClientConnection>(std::move(_socket));
    auto task_runner = std::make_shared<TaskRunner>(_io_service);
    auto session = std::make_shared<ServerSession>(connection, task_runner);
    // Start the session and release it once it has terminated
    session->start() >> then >> [=]() mutable { session.reset(); };
  }

  _accept_next_connection();
}
```

One thing I am pretty sure is odd (and it correlates with the errors we're seeing in CI, somewhat): ` std::make_shared<ClientConnection>(std::move(_socket));` moves the socket into the ClientConnection but then seems to reuse the same, now expired, variable for the next connection? How is this to work?

Also, and this is maybe where my lack of networking knowledge becomes apparent: How is this ever to work with multiple connections in parallel? There is only one socket, ever, right? How can multiple Sessions receive data if there is only one socket? Shouldn't there be a separate socket for each connection?

@toniSta @mrks

**Comment by toniSta** [comment]
There is some boost magic behind it.

>  One thing I am pretty sure is odd (and it correlates with the errors we're seeing in CI, somewhat): std::make_shared<ClientConnection>(std::move(_socket)); moves the socket into the ClientConnection but then seems to reuse the same, now expired, variable for the next connection? How is this to work?

With `std::move(_socket)` you transfer the ownership of the OS socket from the `Server::_acceptor` to the `Server::_socket`. The Boost documentation says about this: 

> Following the move, the moved-from object is in the same state as if constructed using the basic_stream_socket(io_context&) constructor. [see here](https://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/basic_stream_socket/basic_stream_socket/overload6.html)

So in my understanding, you can move the original socket object from server to session as many times as you want. By the way there is no copy constructor for the `basic_stream_socket`.

> Also, and this is maybe where my lack of networking knowledge becomes apparent: How is this ever to work with multiple connections in parallel? There is only one socket, ever, right? How can multiple Sessions receive data if there is only one socket? Shouldn't there be a separate socket for each connection?

So basically you are right: There is only one socket. However, you have to differentiate between server sockets and data sockets. If you open a new connection to an open server socket, a new data socket will be created. This data socket communicates with the client. As far as I know, all the magic happens in the boost acceptor class. In our case, this is done in the following line: `_acceptor.async_accept(_socket, boost::bind(&Server::_start_session, this, boost::asio::placeholders::error));`

**Comment by mrzzzrm** [comment]
This is insanity. Madness. We need torches. Keep boost devs out of my vicinity for a while, please.

Thanks for clearing that up.
